File: 2020/day05/main.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <unordered_set>

using std::cout;
using std::endl;
using std::ifstream;
using std::max_element;
using std::string;
using std::unordered_set;

int main(void) {
    ifstream f("input");
    string line;
    unordered_set<int> seat_ids;
    while (getline(f, line)) {
        replace(line.begin(), line.end(), 'F', '0');
        replace(line.begin(), line.end(), 'B', '1');
        int row = stoi(line.substr(0, 7), NULL, 2);
        replace(line.begin(), line.end(), 'L', '0');
        replace(line.begin(), line.end(), 'R', '1');
        int col = stoi(line.substr(7), NULL, 2);
        seat_ids.insert(row * 8 + col);
    }

    // puzzle 1
    cout << *max_element(seat_ids.begin(), seat_ids.end()) << endl;

    // puzzle 2
    for (int seat_id = 0; seat_id < 1<<10; seat_id++) {
        if (seat_ids.find(seat_id) != seat_ids.end()) {
            continue;
        }
        if (seat_ids.find(seat_id - 1) == seat_ids.end()) {
            continue;
        }
        if (seat_ids.find(seat_id + 1) == seat_ids.end()) {
            continue;
        }
        cout << seat_id << endl;
    }
}

File: 2020/day05/main.py
from typing import Iterator, Tuple


def decode_boarding_pass(boarding_pass: str) -> Tuple[int, int]:
    row = int(boarding_pass[:7].replace('F', '0').replace('B', '1'), 2)
    col = int(boarding_pass[7:].replace('L', '0').replace('R', '1'), 2)
    return row, col


def iterate_seat_ids() -> Iterator[int]:
    with open('input') as f:
        for line in f:
            row, col = decode_boarding_pass(line)
            yield row * 8 + col

def puzzle1() -> None:
    print(max(iterate_seat_ids()))


def puzzle2() -> None:
    seat_ids = set(iterate_seat_ids())
    for missing_seat_id in set(range(2**10)) - seat_ids:
        if {missing_seat_id - 1, missing_seat_id + 1} <= seat_ids:
            print(missing_seat_id)


def main() -> None:
    puzzle1()
    puzzle2()


if __name__ == '__main__':
    main()
